rust/day02/src/main.rs

```diff
@@ -0,0 +1,73 @@
+use std::fs::read_to_string;
+
+// Clone and Copy mean this is a very small object; don't bother with pointers/references ever --
+// just copy it.
+// PartialEq lets us see if x == y
+#[derive(Clone, Copy, PartialEq)]
+enum Choice {
+    Rock,
+    Paper,
+    Scissors,
+}
+// This lets us say Rock instead of Choice::Rock
+use Choice::*;
+
+impl Choice {
+    /// Does this choice beat the other one? `None` if it's a draw
+    fn beats(self, other: Choice) -> Option<bool> {
+        // Draw
+        if self == other {
+            return None;
+        }
+
+        Some(match (self, other) {
+            (Paper, Rock) => true,
+            (Scissors, Paper) => true,
+            (Rock, Scissors) => true,
+            // All other cases, self loses to other
+            _ => false,
+        })
+    }
+
+    fn value(self) -> usize {
+        match self {
+            Rock => 1,
```

Comment:
`match` vs a `HashMap` for these kind of lookup tables doesn't matter too too much. If the set of values is small (< 10 or so?) the performance is basically the same; they'll get compiled down to something similar. The pattern matching is nicer, because you can say things like `case1 | case2 => same value`